server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -1345,14 +1354,138 @@ public static Builder builder(String name, List<Index> indices) {
         return new Builder(name, indices);
     }
 
+    public static Builder builder(String name, DataStreamIndices backingIndices) {
+        return new Builder(name, backingIndices);
+    }
+
     public Builder copy() {
         return new Builder(this);
     }
 
+    public static class DataStreamIndices {
+        private final String namePrefix;
+        private final List<Index> indices;
+        private final boolean rolloverOnWrite;
+        @Nullable
+        private final DataStreamAutoShardingEvent autoShardingEvent;
+        private Set<String> lookup;
+
+        public DataStreamIndices(
+            String namePrefix,
+            List<Index> indices,
+            boolean rolloverOnWrite,
+            DataStreamAutoShardingEvent autoShardingEvent
+        ) {
+            this.namePrefix = namePrefix;
+            this.indices = indices;
```

Comment:
Hmm this is an interesting one. Admittedly, this was not intentional :). I would be inclined to agree with you that doing the copy here makes sense, but my only concern is that this constructor is also called while [parsing](https://github.com/nielsbauman/elasticsearch/blob/d1a4d4975c4337b0ec66dcbe54977d92c86dde8e/server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java#L1066-L1066), which means we'll be copying this every time we deserialize a data stream. However, I don't know how often that happens, and thus whether it would actually be significant. What do you think?

Reply:
I have the same concern. What if we make the constructor non public this way we ensure when it's called it's called within this class where we can argue how it's being used.

Furthermore, we can add a comment that for performance reasons this class expects an immutable list. What do you think?

Reply:
Also as @jbaiera said, we can clean up further in follow up PRs

Reply:
I made the constructor protected as it was being used in some tests.